src/Vm.sol

```diff
@@ -138,12 +138,50 @@ interface VmSafe {
     // struct json = { uint256 a; address b; }
     // If we defined a json struct with the opposite order, meaning placing the address b first, it would try to
     // decode the tuple in that order, and thus fail.
+    //
+    // writeJson
+    //
+    // Let's assume we want to write the following JSON to a file:
+    //
+    // { "boolean": true, "number": 342, "object": { "title": "finally json serialization" } }
+    //
+    // ```
+    //  string memory json1 = "some key";
+    //  vm.serializeBool(json1, "boolean", true);
+    //  vm.serializeBool(json1, "number", uint256(342));
+    //  json2 = "some other key";
+    //  string memory output = vm.serializeString(json2, "title", "finally json serialization");
+    //  vm.serialize(json1, "object", output);
+    //  vm.writeJson(json1, "./output/example.json");
+    // ```
+    //  The critical insight is that every invocation of serialization will return the stringified version of the JSON
+    // up to that point. That means we can construct arbitrary JSON objects and then use the return stringified version
+    // to serialize them as values to another JSON object.
+    //
+    //  json1 and json2 are simply keys used by the backend to keep track of the objects. So vm.serializeJson(json1,..)
+    //  will find the object in-memory that is keyed by "some key".   // writeJson
 
+    function serializeBool(string calldata, string calldata, bool) external returns (string memory);
+    function serializeUint(string calldata, string calldata, uint256) external returns (string memory);
+    function serializeInt(string calldata, string calldata, int256) external returns (string memory);
+    function serializeAddress(string calldata, string calldata, address) external returns (string memory);
+    function serializeBytes32(string calldata, string calldata, bytes32) external returns (string memory);
+    function serializeString(string calldata, string calldata, string calldata) external returns (string memory);
+    function serializeBytes(string calldata, string calldata, bytes calldata) external returns (string memory);
+
+    function serializeBool(string calldata, string calldata, bool[] calldata) external returns (string memory);
+    function serializeUint(string calldata, string calldata, uint256[] calldata) external returns (string memory);
+    function serializeInt(string calldata, string calldata, int256[] calldata) external returns (string memory);
+    function serializeAddress(string calldata, string calldata, address[] calldata) external returns (string memory);
+    function serializeBytes32(string calldata, string calldata, bytes32[] calldata) external returns (string memory);
+    function serializeString(string calldata, string calldata, string[] calldata) external returns (string memory);
+    function serializeBytes(string calldata, string calldata, bytes[] calldata) external returns (string memory);
```

Comment:
This would require a PR to foundry, but since these `serialize*` methods all have difference signatures should they all just be named `serialize`? I'm ok with either, just noticed in `stdJson` they're all overloads called `serialize` so maybe we should be consistent here.

It looks like most cheatcodes just use overloads if function selectors are sufficient to distinguish them, and types are added to the name if it's return values (not part of the selector) that distinguishes them, so perhaps we should make this change

Reply:
@mds1 I had some weird conflicts with how macro was generating the bindings enum, so this was more explicit and solved all issues. Would prefer to keep them like this.